### PR TITLE
[Merged by Bors] - Split bevy_hierarchy out from bevy_transform

### LINKED ISSUE #2758: Should bevy-transform be two crates?
I wrote this issue back in May, but seem to have forgotten to submit it?

Issue writeup below. I apologize if the details turn out to be somewhat out of date.

---

I'm sorry, none of the issue templates seemed to fit. "Feature Request" was the closest, but it didn't quite seem right.

## Question

In `README.md`, you state that one of your design goals is to be *modular*; a user can choose to use only those features they need.
Yet the crate `bevy_transform` seems to bundle two unrelated concerns:
- The concept of a parent-child hierarchy of entities.
- The concept of a `Transform`, roughly a similarity transformation of Euclidean space.

Its description in `Cargo.toml` even admits this — "Provides hierarchy **and** transform functionality for Bevy Engine"! (Emphasis mine.)

Why aren't these two separate crates, with the latter depending on the former?

## Reason

I discovered this issue while exploring the possibility of using Bevy to make games that take place in curved space.
The concept of a parent-child hierarchy is still completely applicable. But `Transform`, at least as currently implemented in Bevy, is not.

## Details of Code Splitting

To ensure that this crate splitting would actually work, I tried it in a clone of the repository. Here are the results.

The crate `bevy_transform` was split into two: `bevy_hierarchy` and `bevy_transform`.

`bevy_hierarchy` contains these things:
- Everything in the modules `components::{children, parent}`.
- Everything in `hierarchy` and its submodules, except for the test in `hierarchy::hierarchy_maintenance`.

It has dependencies `bevy_ecs`, `bevy_reflect`, `bevy_utils`, and `smallvec`.
It does not need `bevy_app` or `bevy_math`.

`bevy-transform` contains everything else:
- Everything in the modules `components::{transform, global_transform}`.
- Everything in the module `transform_propagate_system`.
- Everything in the root module.
- The test from `hierarchy::hierarchy_maintenance`.

It depends on the new `bevy-hierarchy`, as well as `bevy_app`, `bevy_ecs`, `bevy_math`, and `bevy_reflect`.
It does not need `bevy_utils` or `smallvec`.

A few changes were necessary to get `bevy_transform` to compile again:
- Some imports needed to be fixed.
- In three instances, the code `children.0.iter()` started failing, because the field `children.0` is private, and is now defined in a different crate. This was easily fixable; just do `children.iter()` (using the `Deref` impl) instead.

Finally, the dependent crates needed to be adjusted.
- `bevy_scene` depends on `bevy_hierarchy`, but not `bevy_transform`.
- `bevy_gltf` and `bevy_ui` depend on both `bevy_hierarchy` and `bevy_transform`.
- `bevy_internal` needed `bevy_hierarchy` added to its reexports and prelude.
- `bevy_pbr`, `bevy_render`, `bevy_sprite`, and `bevy_text`, despite depending on `bevy_transform`, seem to need no changes.

`cargo --test` passes. I have not done any performance testing.

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- Hierarchy tools are not just used for `Transform`: they are also used for scenes.
- In the future there's interest in using them for other features, such as visiibility inheritance.
- The fact that these tools are found in `bevy_transform` causes a great deal of user and developer confusion
- Fixes #2758.

## Solution

- Split `bevy_transform` into two!
- Make everything work again.

Note that this is a very tightly scoped PR: I *know* there are code quality and docs issues that existed in bevy_transform that I've just moved around. We should fix those in a seperate PR and try to merge this ASAP to reduce the bitrot involved in splitting an entire crate.

## Frustrations

The API around `GlobalTransform` is a mess: we have massive code and docs duplication, no link between the two types and no clear way to extend this to other forms of inheritance.

In the medium-term, I feel pretty strongly that `GlobalTransform` should be replaced by something like `Inherited<Transform>`, which lives in `bevy_hierarchy`:

- avoids code duplication
- makes the inheritance pattern extensible
- links the types at the type-level
- allows us to remove all references to inheritance from `bevy_transform`, making it more useful as a standalone crate and cleaning up its docs

## Additional context

- double-blessed by @cart in https://github.com/bevyengine/bevy/issues/4141#issuecomment-1063592414 and https://github.com/bevyengine/bevy/issues/2758#issuecomment-913810963
- preparation for more advanced / cleaner hierarchy tools: go read https://github.com/bevyengine/rfcs/pull/53 !
- originally attempted by @finegeometer in #2789. It was a great idea, just needed more discussion!

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@cart as part of this PR's merge process, can we get a `A - Hierarchy` label?

File: crates/bevy_hierarchy/.gitignore [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+.idea
+.vscode
+
+
+#Added by cargo
+#
+#already existing elements are commented out
+
+/target
+**/*.rs.bk
```
Comment (mockersf) [inline crates/bevy_hierarchy/.gitignore:24]: this file shouldn't be here

**Review by mockersf — CHANGES_REQUESTED** [review]

File: crates/bevy_hierarchy/src/hierarchy/hierarchy_maintenance_system.rs [hunk 2]
```diff
@@ -78,6 +78,8 @@ mod test {
         world::World,
     };
 
+    use bevy_transform::components::Transform;
```
Comment (james7132) [inline crates/bevy_hierarchy/src/hierarchy/hierarchy_maintenance_system.rs:81]: I feel like this should be rectified. `bevy_hierarchy` should be a dependency of `bevy_transform` not the other way around, even for test dependencies.

File: crates/bevy_ui/src/lib.rs [hunk 3]
```diff
@@ -89,17 +89,17 @@ impl Plugin for UiPlugin {
                 CoreStage::PostUpdate,
                 flex_node_system
                     .label(UiSystem::Flex)
-                    .before(TransformSystem::TransformPropagate),
+                    .before(HierarchySystem::TransformPropagate),
```
Comment (james7132) [inline crates/bevy_ui/src/lib.rs:92]: Mirroring my coments from another file. This seems backwards. TransformPropagate should remain a TransformSystem, and HierarchySystems for maintaining the hierarchy should be separate. Hopefully in the near future, we won't need any HierarchySystems.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Comment by alice-i-cecile** [comment]
bors try

**Comment by IceSentry** [comment]
btw, bevy_hierarchy is not reserved on crates.io. Even if this PR doesn't get merged, we should probably reserve it. 

https://crates.io/crates/bevy_hierarchy

**Review by james7132 — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Comment by cart** [comment]
> The API around GlobalTransform is a mess: we have massive code and docs duplication, no link between the two types and no clear way to extend this to other forms of inheritance. In the medium-term, I feel pretty strongly that GlobalTransform should be replaced by something like Inherited<Transform>, which lives in bevy_hierarchy

I object to the framing that a fully separate GlobalTransform component is a "mess". I personally think a named (and fully separate) GlobalTransform type is preferable to a more abstracted "component inheritance system". GlobalTransform having a separate type (and name) makes the context and "space" (as in "world space") of the component clear. It also leaves the door open to things like "adding a `GlobalTransform::ignore_parent` field".

On the other hand, `Inherited<Transform>` could literally mean anything (and _would_ mean different things, depending on the context of `Inherited + ComponentType`). Of course, the value of this design is up for debate (and should be discussed in a different place), but by including "highly opinionated and negatively weighted" thoughts on what we already have, I feel compelled to reply here.

**Comment by cart** [comment]
(but I do agree that it would be nice to cut down on code duplication ... impl macros or Deref are also options)

File: crates/bevy_transform/src/lib.rs [hunk 4]
```diff
@@ -81,46 +77,33 @@ impl From<Transform> for TransformBundle {
         Self::from_transform(transform)
     }
 }
-/// The base plugin for handling [`Transform`] components
-#[derive(Default)]
-pub struct TransformPlugin;
-
-/// Label enum for the types of systems relating to transform
+/// Label enum for the systems relating to transform propagaion
```
Comment (cart) [inline crates/bevy_transform/src/lib.rs:80]:
```suggestion
/// Label enum for the systems relating to transform propagation
```

**Review by cart — APPROVED** [review]
Looks good to me. Just one baby nit.

**Comment by alice-i-cecile** [comment]
>  but by including "highly opinionated and negatively weighted" thoughts on what we already have, I feel compelled to reply here.

Fair; this wasn't really the place or tone for it. I'm sorry for that @cart. I was frustrated about the code duplication, but we can debate the relative merits of that design somewhere else.

**Comment by cart** [comment]
> Fair; this wasn't really the place or tone for it. I'm sorry for that @cart. I was frustrated about the code duplication, but we can debate the relative merits of that design somewhere else.

No worries. I almost didn't say anything because its definitely ok to think something is a mess (although I try to avoid using that language to avoid hurting peoples' feelings ... and to account for the fact that I might find out I'm wrong about that assessment. When trying to get that point across, I think its better to just outline the specific facts of the situation rather than ascribe generic negative labels to things). I mainly brought it up because I'd rather leave "opinionated discussions and pronouncements about the value of a thing" to issues scoped to that topic. I selfishly don't want to defend my design choices in 10 different places :)

**Pull request closed without merging**